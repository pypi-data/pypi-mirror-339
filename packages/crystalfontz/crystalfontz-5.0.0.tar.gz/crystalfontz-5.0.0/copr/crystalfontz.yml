apiVersion: coprctl/v1alpha1
kind: package-scm
metadata:
  name: crystalfontz
spec:
  auto_rebuild: true
  ownername: jfhbrook
  projectname: joshiverse
  clone-url: https://github.com/jfhbrook/crystalfontz
  commit: null
  spec: crystalfontz.spec
  type: git
  method: tito
