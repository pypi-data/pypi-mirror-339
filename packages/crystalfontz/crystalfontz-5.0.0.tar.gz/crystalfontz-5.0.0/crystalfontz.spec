Name: crystalfontz
Version: 5.0.0
Release: 1
License: MPL-2.0
Summary: Serial client and Linux service for Crystalfontz LCD displays

URL: https://github.com/jfhbrook/crystalfontz
Source0: %{name}-%{version}.tar.gz
BuildArch: noarch

Requires: python-crystalfontz
Requires: python-sdbus

%description


%prep
%autosetup


%build
tar -xzf %{SOURCE0}


%install
mkdir -p %{buildroot}%{_prefix}/lib/systemd/system
mkdir -p %{buildroot}%{_bindir}
install -p -D -m 0644 systemd/crystalfontz.service %{buildroot}%{_prefix}/lib/systemd/system/crystalfontz.service
install -p -D -m 0644 dbus/org.jfhbrook.crystalfontz.conf %{buildroot}%{_prefix}/share/dbus-1/system.d/org.jfhbrook.crystalfontz.conf
install -p -m 755 bin/crystalfontz-dbus %{buildroot}%{_bindir}/crystalfontz

%check


%files
%{_prefix}/lib/systemd/system/crystalfontz.service
%{_prefix}/share/dbus-1/system.d/org.jfhbrook.crystalfontz.conf
%{_bindir}/crystalfontz

%changelog
* Tue Apr 08 2025 Josh Holbrook <josh.holbrook@gmail.com> 5.0.0-1
- API Changes:
  - **NEW:** `Receiver` class
    - A subclass of `asyncio.Queue`
  - `Client`:
    - **BREAKING:** `subscribe` and `unsubscribe` methods use new `Receiver` class
    - Emit unmatched exceptions on expecting receivers instead of resolving `client.closed`
    - `detect_baud_rate` exposes `timeout` and `retry_times` arguments
    - Document `detect_baud_rate`
  - **BREAKING:** `ClientProtocol`/`EffectClient`
    - `crystalfontz.protocol.ClientProtocol` type has been replaced by `crystalfontz.effects.EffectClient`
    - `EffectClient` enforces a smaller API than `ClientProtocol` did previously
  - **BREAKING:** `Response`:
    - `Response.from_bytes` accepts bytes as from packets, rather than `__init__`
    - `Response.__init__` accepts properties as arguments
  - **BREAKING:** `SpecialCharacter` API:
    - Rename `as_bytes` method to `to_bytes`
    - Store pixels as `List[List[bool]]` instead of `List[List[int]]`
  - `KeyState` includes `keypress: KeyPress` attribute
- CLI changes:
  - **BREAKING:** CLI now invoked through `python3 -m crystalfontz`
    - Optional unpackaged `crystalfontz` entry point at `./bin/crystalfontz`
  - Use `configurence` library
  - Respect `CRYSTALFONTZ_CONFIG_FILE` environment variable
  - Improve error reporting for timeouts
  - Client now respects CLI settings
    - Cause was the `Client` constructor being called twice
  - Additional commands accept bytes as arguments
    - `python3 -m crystalfontz line 1`
    - `python3 -m crystalfontz line 2`
    - `python3 -m crystalfontz send`
  - Help for `python3 -m crystalfontz listen` `--for` option
- **NEW:** Dbus support:
  - `crystalfontz.dbus.DbusInterface` dbus Interface class, implementing most commands
  - `crystalfontz.dbus.DbusClient` dbus client class
  - `crystalfontz.dbus.domain` API for mapping domain objects to dbus types
  - `python3 -m crystalfontz.dbus.service` dbus service CLI
    - Optional unpackaged `crystalfontz.dbus.service` entry point at `./bin/crystalfontz-service`
  - `python3 -m crystalfontz.dbus.client` dbus client CLI
    - Optional unpackaged `crystalfontz.dbus.client` entry point at `./bin/crystalfontz-dbus`
- Packaging and Releases:
  - `python-crystalfontz` COPR package spec
    - This has been moved from <https://github.com/jfhbrook/public>
  - **NEW:** `crystalfontz` COPR package
    - Depends on `python-crystalfontz` COPR package
    - Includes systemd unit for `python3 -m crystalfontz.dbus.service`
    - **BREAKING:** Includes shim bin `crystalfontz` -> `python3 -m crystalfontz.dbus.client`
  - Tito based release tagging
  - GitHub release
  - Improved PyPI classifiers
  - **BREAKING:** Release under MPL-2.0 license
- Integration tests:
  - Use `python-gak` plugin
  - Use snapshots
  - Leverage a config file at `./tests/fixtures/crystalfontz.yaml`
    - Can be overridden with `CRYSTALFONTZ_CONFIG_FILE` environment variable


