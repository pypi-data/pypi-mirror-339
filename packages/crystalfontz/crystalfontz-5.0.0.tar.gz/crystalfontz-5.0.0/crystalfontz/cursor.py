from enum import Enum


class CursorStyle(Enum):
    """
    A cursor style, as set with command 12 (0x0C): Set LCD Cursor Style.

    - **NONE**:  No cursor.
    - **BLINKING_BLOCK**: Blinking block cursor.
    - **STATIC_UNDERSCORE**: Static underscore cursor.
    - **BLINKING_UNDERSCORE**: Blinking underscore cursor. On the CFA633, this
      represents a blinking block plus an underscore.
    """

    NONE = 0
    BLINKING_BLOCK = 1
    STATIC_UNDERSCORE = 2
    BLINKING_UNDERSCORE = 3
