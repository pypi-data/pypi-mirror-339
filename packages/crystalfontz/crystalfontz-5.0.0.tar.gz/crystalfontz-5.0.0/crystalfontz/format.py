import base64
from typing import Literal

OutputMode = Literal["text"] | Literal["json"]


def format_bytes(buffer: bytes) -> str:
    return str(buffer)[2:-1]


def format_json_bytes(buffer: bytes) -> str:
    return base64.b64encode(buffer).decode("utf-8")
