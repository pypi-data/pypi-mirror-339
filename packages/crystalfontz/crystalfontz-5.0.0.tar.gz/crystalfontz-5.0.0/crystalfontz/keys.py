from dataclasses import asdict, dataclass
from enum import Enum
from typing import List, Literal, Self, Tuple, Type

KeyPress = (
    Literal[0x01]
    | Literal[0x02]
    | Literal[0x04]
    | Literal[0x08]
    | Literal[0x10]
    | Literal[0x20]
)

KP_UP: KeyPress = 0x01
KP_ENTER: KeyPress = 0x02
KP_EXIT: KeyPress = 0x04
KP_LEFT: KeyPress = 0x08
KP_RIGHT: KeyPress = 0x10
KP_DOWN: KeyPress = 0x20


def keypress_repr(keypress: KeyPress) -> str:
    return {
        KP_UP: "KP_UP",
        KP_ENTER: "KP_ENTER",
        KP_EXIT: "KP_EXIT",
        KP_LEFT: "KP_LEFT",
        KP_RIGHT: "KP_RIGHT",
        KP_DOWN: "KP_DOWN",
    }[keypress]


@dataclass
class KeyState:
    """
    A key's state.

    Attributes:
        pressed (bool): When True, the key is currently pressed.
        pressed_since (bool): When True, the key has been pressed since the last poll.
        released_since (bool): When True, the key has been released since the last
                               poll.
    """

    keypress: KeyPress
    pressed: bool
    pressed_since: bool
    released_since: bool

    @classmethod
    def from_bytes(cls: Type[Self], state: bytes, keypress: KeyPress) -> Self:
        pressed = state[0]
        pressed_since = state[1]
        released_since = state[2]

        return cls(
            keypress=keypress,
            pressed=bool(pressed & keypress),
            pressed_since=bool(pressed_since & keypress),
            released_since=bool(released_since & keypress),
        )

    def to_bytes(self: Self) -> Tuple[int, int, int]:
        pressed = self.keypress if self.pressed else 0x00
        pressed_since = self.keypress if self.pressed_since else 0x00
        released_since = self.keypress if self.released_since else 0x00

        return (pressed, pressed_since, released_since)


@dataclass
class KeyStates:
    """
    The state of all keys.

    Attributes:
        up: The state of the "up" key.
        enter: The state of the "enter" key.
        exit: The state of the "exit" key.
        left: The state of the "left" key.
        right: The state of the "right" key.
        down: The state of the "down" key.
    """

    up: KeyState
    enter: KeyState
    exit: KeyState
    left: KeyState
    right: KeyState
    down: KeyState

    @classmethod
    def from_bytes(cls: Type[Self], state: bytes) -> Self:
        return cls(
            up=KeyState.from_bytes(state, KP_UP),
            enter=KeyState.from_bytes(state, KP_ENTER),
            exit=KeyState.from_bytes(state, KP_EXIT),
            left=KeyState.from_bytes(state, KP_LEFT),
            right=KeyState.from_bytes(state, KP_RIGHT),
            down=KeyState.from_bytes(state, KP_DOWN),
        )

    def to_bytes(self: Self) -> bytes:
        pressed = 0x00
        pressed_since = 0x00
        released_since = 0x00

        for state in [
            self.up,
            self.enter,
            self.exit,
            self.left,
            self.right,
            self.down,
        ]:
            p, p_s, r_s = state.to_bytes()
            pressed = pressed ^ p
            pressed_since = pressed_since ^ p_s
            released_since = released_since ^ r_s

        return bytes([pressed, pressed_since, released_since])

    def __repr__(self: Self) -> str:
        repr_ = ""
        for name, state in asdict(self).items():
            st = ", ".join(
                [
                    (
                        f"{n}={'yes' if s else 'no'}"
                        if isinstance(s, bool)
                        else f"{n}={keypress_repr(s)}"
                    )
                    for n, s in state.items()
                ]
            )
            repr_ += f"{name}: {st}\n"
        return repr_[0:-1]


class KeyActivity(Enum):
    """
    A key activity. This is either a "press" event or a "release" event for a
    given key.
    """

    KEY_UP_PRESS = 1
    KEY_DOWN_PRESS = 2
    KEY_LEFT_PRESS = 3
    KEY_RIGHT_PRESS = 4
    KEY_ENTER_PRESS = 5
    KEY_EXIT_PRESS = 6
    KEY_UP_RELEASE = 7
    KEY_DOWN_RELEASE = 8
    KEY_LEFT_RELEASE = 9
    KEY_RIGHT_RELEASE = 10
    KEY_ENTER_RELEASE = 11
    KEY_EXIT_RELEASE = 12

    @classmethod
    def from_byte(cls: Type[Self], activity: int) -> "KeyActivity":
        return KEY_ACTIVITIES[activity - 1]

    @classmethod
    def from_bytes(cls: Type[Self], activity: bytes) -> "KeyActivity":
        return cls.from_byte(activity[0])

    def to_byte(self: Self) -> int:
        return self.value


KEY_ACTIVITIES: List[KeyActivity] = [
    KeyActivity.KEY_UP_PRESS,
    KeyActivity.KEY_DOWN_PRESS,
    KeyActivity.KEY_LEFT_PRESS,
    KeyActivity.KEY_RIGHT_PRESS,
    KeyActivity.KEY_ENTER_PRESS,
    KeyActivity.KEY_EXIT_PRESS,
    KeyActivity.KEY_UP_RELEASE,
    KeyActivity.KEY_DOWN_RELEASE,
    KeyActivity.KEY_LEFT_RELEASE,
    KeyActivity.KEY_RIGHT_RELEASE,
    KeyActivity.KEY_ENTER_RELEASE,
    KeyActivity.KEY_EXIT_RELEASE,
]
