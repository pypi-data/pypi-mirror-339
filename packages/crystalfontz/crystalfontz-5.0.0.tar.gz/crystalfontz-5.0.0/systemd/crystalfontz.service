[Unit]
Description=Crystalfontz LCD
Requires=dbus.socket
After=dbus.socket

[Service]
Type=dbus
BusName=org.jfhbrook.crystalfontz
ExecStart=/usr/bin/python3 -m crystalfontz.dbus.service --system
Restart=on-failure

[Install]
WantedBy=multi-user.target
