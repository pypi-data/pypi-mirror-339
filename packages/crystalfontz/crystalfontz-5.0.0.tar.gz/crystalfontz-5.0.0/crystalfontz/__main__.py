from crystalfontz.cli import main

main()
