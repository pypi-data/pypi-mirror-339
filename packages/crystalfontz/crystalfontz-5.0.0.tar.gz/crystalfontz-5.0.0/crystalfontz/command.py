from abc import ABC, abstractmethod
from functools import reduce
from typing import Iterable, Optional, Self, Set
import warnings

from crystalfontz.atx import AtxPowerSwitchFunctionalitySettings
from crystalfontz.baud import BaudRate
from crystalfontz.character import SpecialCharacter
from crystalfontz.cursor import CursorStyle
from crystalfontz.device import Device
from crystalfontz.gpio import GpioSettings
from crystalfontz.keys import KeyPress
from crystalfontz.lcd import LcdRegister
from crystalfontz.packet import Packet
from crystalfontz.temperature import pack_temperature_settings, TemperatureDisplayItem

SET_LINE_WARNING_TEMPLATE = (
    "Command {code} ({code:02x}): {name} is deprecated"
    " in favor of command 31 (0x1F): Send Data to LCD."
)


class Command(ABC):
    @abstractmethod
    def to_packet(self: Self) -> Packet:
        raise NotImplementedError("to_packet")


class Ping(Command):
    command: int = 0x00

    def __init__(self: Self, payload: bytes) -> None:
        if len(payload) > 16:
            raise ValueError(f"Payload length {len(payload)} > 16")
        self.payload: bytes = payload

    def to_packet(self) -> Packet:
        return (self.command, self.payload)


class GetVersions(Command):
    command: int = 0x01

    def to_packet(self: Self) -> Packet:
        return (self.command, b"")


class WriteUserFlashArea(Command):
    command: int = 0x02

    def __init__(self: Self, data: bytes) -> None:
        self.data: bytes = data

    def to_packet(self: Self) -> Packet:
        return (self.command, self.data)


class ReadUserFlashArea(Command):
    command: int = 0x03

    def to_packet(self: Self) -> Packet:
        return (self.command, b"")


class StoreBootState(Command):
    command: int = 0x04

    def to_packet(self: Self) -> Packet:
        return (self.command, b"")


class PowerCommand(Command):
    command: int = 0x05

    pass


class RebootLCD(PowerCommand):
    def to_packet(self: Self) -> Packet:
        return (self.command, bytes([8, 18, 99]))


class ResetHost(PowerCommand):
    def to_packet(self: Self) -> Packet:
        return (self.command, bytes([12, 28, 97]))


class ShutdownHost(PowerCommand):
    def to_packet(self: Self) -> Packet:
        return (self.command, bytes([3, 11, 95]))


class ClearScreen(Command):
    command: int = 0x06

    def to_packet(self: Self) -> Packet:
        return (self.command, b"")


class SetLine1(Command):
    command: int = 0x07

    def __init__(self: Self, line: str | bytes, device: Device) -> None:
        warnings.warn(
            SET_LINE_WARNING_TEMPLATE.format(
                code=0x07, name="Set LCD Contents, Line 1"
            ),
            DeprecationWarning,
        )
        buffer: bytes = (
            device.character_rom.encode(line) if isinstance(line, str) else line
        )

        if len(buffer) > device.columns:
            raise ValueError(f"Line length {len(buffer)} longer than {device.columns}")

        self.line = buffer.ljust(device.columns, b" ")

    def to_packet(self: Self) -> Packet:
        return (self.command, self.line)


class SetLine2(Command):
    command: int = 0x08

    def __init__(self: Self, line: str | bytes, device: Device) -> None:
        warnings.warn(
            SET_LINE_WARNING_TEMPLATE.format(
                code=0x08, name="Set LCD Contents, Line 2"
            ),
            DeprecationWarning,
        )
        buffer: bytes = (
            device.character_rom.encode(line) if isinstance(line, str) else line
        )

        if len(buffer) > device.columns:
            raise ValueError(f"Line length {len(buffer)} longer than {device.columns}")

        self.line: bytes = buffer.ljust(device.columns, b" ")

    def to_packet(self: Self) -> Packet:
        return (self.command, self.line)


class SetSpecialCharacterData(Command):
    command: int = 0x09

    def __init__(
        self: Self, index: int, character: SpecialCharacter, device: Device
    ) -> None:
        device.character_rom.validate_special_character_index(index)
        self.index: int = index
        self.character: bytes = character.to_bytes(device)

    def to_packet(self: Self) -> Packet:
        data = self.index.to_bytes(1, "big") + self.character
        assert len(data) == 9
        return (self.command, data)


class ReadLcdMemory(Command):
    command: int = 0x0A

    def __init__(self: Self, address: int) -> None:
        # Address is native to the LCD controller. On the CFA533 h1.4 u1v2,
        # they are:
        #
        #     [0x40, 0x7F] -> CGRAM
        #     [0x80, 0x93] -> DDRAM, line 1
        #     [0xC0, 0xD3] -> DDRAM, line 2
        #
        # These are likely specific to the model and revision. Rather than
        # attempt to validate these at the Device level, we simply assert
        # that the address is a valid byte, and leave choosing sensible
        # addresses as an exercise for the user.
        if not (0 < address < 255):
            raise ValueError(f"Address {address} is invalid")
        self.address: bytes = address.to_bytes(1, "big")

    def to_packet(self: Self) -> Packet:
        return (self.command, self.address)


class SetCursorPosition(Command):
    command: int = 0x0B

    def __init__(self: Self, row: int, column: int, device: Device) -> None:
        if column < 0:
            raise ValueError(f"Column {column} < 0")
        elif column >= device.columns:
            raise ValueError(f"Column {column} >= {device.columns}")
        if row < 0:
            raise ValueError(f"Row {row} < 0")
        elif row >= device.lines:
            raise ValueError(f"Row {row} >= {device.lines}")

        self.row = row
        self.column = column

    def to_packet(self: Self) -> Packet:
        return (
            self.command,
            self.column.to_bytes(1, "big") + self.row.to_bytes(1, "big"),
        )


class SetCursorStyle(Command):
    command: int = 0x0C

    def __init__(self, style: CursorStyle) -> None:
        self.style: bytes = style.value.to_bytes(1, "big")

    def to_packet(self: Self) -> Packet:
        return (self.command, self.style)


class SetContrast(Command):
    command: int = 0x0D

    def __init__(self: Self, contrast: float, device: Device) -> None:
        self.contrast = device.contrast(contrast)

    def to_packet(self: Self) -> Packet:
        return (self.command, self.contrast)


class SetBacklight(Command):
    command: int = 0x0E

    def __init__(
        self: Self,
        lcd_brightness: float,
        keypad_brightness: Optional[float],
        device: Device,
    ) -> None:
        self.brightness = device.brightness(lcd_brightness, keypad_brightness)

    def to_packet(self: Self) -> Packet:
        return (self.command, self.brightness)


# 0x0F-0x11 are reserved


class ReadDowDeviceInformation(Command):
    command: int = 0x12

    def __init__(self: Self, index: int) -> None:
        # TODO: CFA533 says this must be < 31.
        self.index: int = index

    def to_packet(self: Self) -> Packet:
        return (self.command, self.index.to_bytes(1, "big"))


class SetupTemperatureReporting(Command):
    command: int = 0x13

    def __init__(self: Self, enabled: Iterable[int], device: Device) -> None:
        self.settings = pack_temperature_settings(enabled, device)

    def to_packet(self: Self) -> Packet:
        return (self.command, self.settings)


class DowTransaction(Command):
    command: int = 0x14

    def __init__(
        self: Self, index: int, bytes_to_read: int, data_to_write: bytes
    ) -> None:
        self.index = index
        if not (0 <= bytes_to_read <= 14):
            raise ValueError("bytes_to_read out of range")
        if len(data_to_write) > 14:
            raise ValueError(
                "data_to_write has {len(data_to_write)} bytes, but only "
                "14 may be written"
            )
        self.bytes_to_read = bytes_to_read
        self.data_to_write = data_to_write

    def to_packet(self: Self) -> Packet:
        return (
            self.command,
            self.index.to_bytes(1, "big")
            + self.bytes_to_read.to_bytes(1, "big")
            + self.data_to_write,
        )


class SetupLiveTemperatureDisplay(Command):
    command: int = 0x15

    def __init__(
        self: Self, slot: int, item: Optional[TemperatureDisplayItem], device: Device
    ) -> None:
        if not (0 <= slot < 8):
            raise ValueError("Slot must be between 0 and 7")
        self.slot: int = slot
        self.item: bytes = TemperatureDisplayItem.to_bytes(item, device)

    def to_packet(self: Self) -> Packet:
        return (self.command, self.slot.to_bytes(1, "big") + self.item)


class SendCommandToLcdController(Command):
    command: int = 0x16

    def __init__(self: Self, location: LcdRegister, data: int | bytes) -> None:
        byte: bytes = data.to_bytes(1, "big") if isinstance(data, int) else data

        if len(byte) != 1:
            raise ValueError("May send one byte to LCD controller")

        self.location = location
        self.byte = byte

    def to_packet(self: Self) -> Packet:
        return (self.command, self.location.value.to_bytes(1, "big") + self.byte)


def _key_mask(keypresses: Set[KeyPress]) -> int:
    return reduce(lambda mask, keypress: mask ^ keypress, keypresses, 0x00)


class ConfigureKeyReporting(Command):
    command: int = 0x17

    def __init__(
        self: Self, when_pressed: Set[KeyPress], when_released: Set[KeyPress]
    ) -> None:
        self.when_pressed: int = _key_mask(when_pressed)
        self.when_released: int = _key_mask(when_released)

    def to_packet(self: Self) -> Packet:
        return (
            self.command,
            self.when_pressed.to_bytes(1, "big")
            + self.when_released.to_bytes(1, "big"),
        )


class PollKeypad(Command):
    command: int = 0x18

    def to_packet(self: Self) -> Packet:
        return (self.command, b"")


# 0x19-0x1B are reserved


class SetAtxPowerSwitchFunctionality(Command):
    command: int = 0x1C

    def __init__(self: Self, settings: AtxPowerSwitchFunctionalitySettings) -> None:
        self.settings = settings

    def to_packet(self: Self) -> Packet:
        return (self.command, self.settings.to_bytes())


class ConfigureWatchdog(Command):
    command: int = 0x1D

    def __init__(self: Self, timeout_seconds: int) -> None:
        if not (0 <= timeout_seconds <= 255):
            raise ValueError("Watchdog timeout must be between 0 and 255 seconds")
        self.timeout_seconds = timeout_seconds

    def to_packet(self: Self) -> Packet:
        return (self.command, self.timeout_seconds.to_bytes(1, "big"))


class ReadStatus(Command):
    command: int = 0x1E

    def to_packet(self: Self) -> Packet:
        return (self.command, b"")


class SendData(Command):
    command: int = 0x1F

    def __init__(
        self: Self, row: int, column: int, text: str | bytes, device: Device
    ) -> None:
        if not (0 <= row < device.lines):
            raise ValueError(f"{row} is not a valid row")
        if not (0 <= column < device.columns):
            raise ValueError(f"{column} is not a valid column")

        buffer: bytes = (
            device.character_rom.encode(text) if isinstance(text, str) else text
        )

        if len(buffer) > device.columns:
            raise ValueError(f"Text length {len(buffer)} longer than {device.columns}")

        self.row: int = row
        self.column: int = column
        self.text: bytes = buffer

    def to_packet(self: Self) -> Packet:
        return (
            self.command,
            self.column.to_bytes(1, "big") + self.row.to_bytes(1, "big") + self.text,
        )


# 0x20 is reserved for CFA631 key legends


class SetBaudRate(Command):
    command: int = 0x21

    def __init__(self: Self, rate: BaudRate) -> None:
        self.baud_rate: int = 0
        if rate == 115200:
            self.baud_rate = 1
        elif rate != 19200:
            raise ValueError(f"Unsupported baud rate {rate}")

    def to_packet(self: Self) -> Packet:
        return (self.command, self.baud_rate.to_bytes(1, "big"))


class SetGpio(Command):
    command: int = 0x22

    def __init__(
        self: Self,
        index: int,
        output_state: int,
        settings: Optional[GpioSettings] = None,
    ) -> None:
        if not (0 <= index < 255):
            raise ValueError(f"Invalid index {index}")

        if not (0 <= output_state <= 100):
            raise ValueError("Output state should be between 0 and 100")

        self.index: int = index
        self.output_state: int = output_state
        self.settings: Optional[GpioSettings] = settings

    def to_packet(self: Self) -> Packet:
        data: bytes = self.index.to_bytes(1, "big") + self.output_state.to_bytes(
            1, "big"
        )

        if self.settings is not None:
            data += self.settings.to_bytes()

        return (self.command, data)


class ReadGpio(Command):
    command: int = 0x23

    def __init__(self: Self, index: int) -> None:
        self.index: int = index

    def to_packet(self: Self) -> Packet:
        return (self.command, self.index.to_bytes(1, "big"))
