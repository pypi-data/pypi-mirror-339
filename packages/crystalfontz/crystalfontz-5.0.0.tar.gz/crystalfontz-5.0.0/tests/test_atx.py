import pytest

from crystalfontz.atx import AtxPowerSwitchFunction, AtxPowerSwitchFunctionalitySettings


@pytest.mark.parametrize(
    "settings",
    [
        AtxPowerSwitchFunctionalitySettings(
            functions={AtxPowerSwitchFunction.KEYPAD_RESET},
            auto_polarity=True,
            reset_invert=False,
            power_invert=False,
            power_pulse_length_seconds=1.0,
        )
    ],
)
def test_atx_settings_to_from_bytes(
    settings: AtxPowerSwitchFunctionalitySettings, snapshot
) -> None:
    as_bytes = settings.to_bytes()

    assert as_bytes == snapshot

    from_bytes = AtxPowerSwitchFunctionalitySettings.from_bytes(as_bytes)

    assert from_bytes == settings
