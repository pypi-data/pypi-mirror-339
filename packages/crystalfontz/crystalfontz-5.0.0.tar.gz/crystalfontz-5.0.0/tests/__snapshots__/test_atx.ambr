# serializer version: 1
# name: test_atx_settings_to_from_bytes[settings0]
  b'! '
# ---
